src/bin/ipa_bench/gen_events.rs

```diff
@@ -0,0 +1,481 @@
+use super::sample::Sample;
+use log::{debug, info, trace};
+use rand::SeedableRng;
+use rand::{CryptoRng, Rng, RngCore};
+use rand_chacha::ChaCha20Rng;
+use rand_distr::num_traits::ToPrimitive;
+use rand_distr::{Bernoulli, Distribution};
+use raw_ipa::helpers::models::{
+    Event as EEvent, SecretSharable, SecretShare, SourceEvent as ESourceEvent,
+    TriggerEvent as ETriggerEvent,
+};
+use serde::{Deserialize, Serialize};
+use std::io;
+use std::time::Duration;
+
+const DAYS_IN_EPOCH: u64 = 7;
+
+#[derive(Clone)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Event {
+    // For this tool, we'll fix the length of a matchkey to u64
+    pub matchkeys: Vec<u64>,
+    pub epoch: u8,
+    pub timestamp: u32,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct SourceEvent {
+    pub event: Event,
+    pub breakdown_key: String,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct TriggerEvent {
+    pub event: Event,
+    pub value: u32,
+    pub zkp: String,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub enum EventType {
+    // Source event in clear
+    S(SourceEvent),
+    // Trigger event in clear
+    T(TriggerEvent),
+    // Source event in cipher if --secret-share option is enabled
+    ES(ESourceEvent),
+    // Trigger event in cipher if --secret-share option is enabled
+    ET(ETriggerEvent),
+}
+
+struct GenEventParams {
+    devices: u8,
+    impressions: u8,
+    conversions: u8,
+    epoch: u8,
+    breakdown_key: String,
+}
+
+// TODO: Currently, users are mutually exclusive in each ad loop (i.e. User A in ad X will never appear in other ads).
+// We need to generate events from same users across ads (but how often should a user appear in different ads?)
+// "Ads" doesn't mean FB's L3 ads. It could be ads from different businesses.
+
+pub fn generate_events<W: io::Write>(
+    sample: &Sample,
+    total_count: u32,
+    epoch: u8,
+    secret_share: bool,
+    seed: &Option<u64>,
+    out: &mut W,
+) -> (u32, u32) {
+    let mut rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+    debug!("seed: {:?}", rng.get_seed());
+
+    // Separate RNG for generating secret shares
+    let mut ss_rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
```

Comment:
Use rand::rngs::StdRng here.  No point in pulling in an explicit dependency.  That means that you don't get seed_from_u64, so you need to change how you invoke it.  You might need to print the seed in hex (and do some hex decoding) in that case.

Note that it is possible to have the hex conversion done by structopt: https://github.com/martinthomson/ohttp/blob/e27ee1af5c698091105c43cf86028ec86b2cbdb7/ohttp-client/src/main.rs#L12-L26